lib/mixins/property-effects.js

```diff
@@ -2682,14 +2694,23 @@ export const PropertyEffects = dedupingMixin(superClass => {
         // and link into list of templates if necessary
         templateInfo = /** @type {!TemplateInfo} */(Object.create(templateInfo));
         templateInfo.wasPreBound = wasPreBound;
-        if (!wasPreBound && this.__templateInfo) {
-          let last = this.__templateInfoLast || this.__templateInfo;
-          this.__templateInfoLast = last.nextTemplateInfo = templateInfo;
-          templateInfo.previousTemplateInfo = last;
-          return templateInfo;
+        if (wasPreBound || !this.__templateInfo) {
```

Comment:
This method is called in a few different ways:

* During finalization (`instanceBinding==false`, `wasPreBound==false`):
  * `_bindTemplate(false)` called directly during finalization - parses the template (for the first time), and then assigns that _prototypical_ template info to `__templateInfo` _on the prototype_; note in this case `wasPreBound` is false; this is the first time we're binding it, thus we create accessors.
* During first stamping (`instanceBinding==true`, `wasPreBound==true`):
  * `_stampTemplate` calls `_bindTemplate(true)` - the `templateInfo` returned matches the prototypical one, and so this is `wasPreBound == true` state; thus we _skip_ creating accessors, but _do_ create a sub-instance of the template info to serve as the start of our linked list (needs to be an instance, not the prototypical one, so that we can chain runtime-stamped template infos off of it). Note that this shadows the prototypical `__templateInfo`, such that subsequent stampings of the same template will not be considered "prebound"
* During subsequent runtime stamping (`instanceBinding==true`, `wasPreBound==false`):
  * `_stampTemplate` calls `_bindTemplate(true)` - here `templateInfo` is guaranteed to _not_ match the prototypical one, because it was either a different template altogether, or even if it was the same template, the step above created a sub-instance of the info; in this case `wasPreBound == false`, so we _do_ create accessors, _and_ link a sub-instance into the linked list.

So in short...
* `instanceBinding` determines whether to link a sub-instance of the template info into the linked list (at the root `__templateInfo` if it doesn't exist yet), or set just set it onto the prototype
* `wasPreBound` determines whether we create accessors on the instance or not

So...
* No `wasPreBound` can't return early since it still needs to put a sub-instance of its `templateInfo` on as the root of the linked list.
* And no, `instanceBinding` determines whether you set `__templateInfo` to `templateInfo` or `Object.create(templateInfo)`

However, I think perhaps one improvement to make it less confusing might be to not use the same prototypical property name as the instance property name, since they're basically used differently.  I'll change the prototypical one to `__preBoundTemplateInfo`, whereas the instance-time linked list would stay `__templateInfo`